llama.cpp

```diff
@@ -17016,6 +17016,20 @@ static int32_t llama_chat_apply_template_internal(
         if (add_ass) {
             ss << "<|START_OF_TURN_TOKEN|><|CHATBOT_TOKEN|>";
         }
+    } else if (tmpl == "llama3" || (tmpl.find("<|start_header_id|>") != std::string::npos && tmpl.find("<|end_header_id|>") != std::string::npos)) {
+        // Llama 3
+        bool first_message = true; // add BOS at the beginning
+        for (auto message : chat) {
+            std::string role(message->role);
+            if (first_message) {
+                ss << "<|begin_of_text|>";
```

Comment:
You don't need to include `<|begin_of_text|>` if it is already the BOS token (which is the case [here](https://huggingface.co/unsloth/llama-3-8b-Instruct/blob/main/tokenizer_config.json#L2052)). BOS token will be added by tokenizer.

Reply:
Okay thanks! Will do.